haystack/nodes/other/shaper.py

```diff
@@ -0,0 +1,419 @@
+    (query, documents etc.) of a pipeline and pass the new/modified variables further down the pipeline.
+
+    Using YAML configuration Shaper component is initialized with functions to invoke on pipeline invocation
+    context.
+
+    For example, in the YAML snippet below:
+    ```yaml
+        version: ignore
+        components:
+        - name: shaper
+          params:
+            inputs:
+                query:
+                  func: expand
+                  output: questions
+          type: Shaper
+        pipelines:
+          - name: query
+            nodes:
+              - name: shaper
+                inputs:
+                  - Query
+    ```
+    Shaper component is initialized with a directive to invoke function expand on the variable query and to store
+    the result in the invocation context variable questions. All other invocation context variables are passed down
+    the pipeline as is.
+
+    Shaper is especially useful in the context of pipelines with PromptNode(s) where we need to modify the invocation
+    context to match the templates of PromptNodes.
+
+    Multiple Shaper components can be used in a pipeline to modify the invocation context as needed.
+
+    The following functions are supported by Shaper:
+
+    - len: parameters are target:Any; returns int; built-in len function returning the length of the input variable
+    - expand: parameters are [expand_target:Any, size:int]; returns List[Any]; it returns a list of specified size
+        for the input variable
+    - concat: parameters are [texts:List[str], delimiter:str]; returns str, concatenates texts with the specified
+        delimiter
+    - concat_docs: parameters are [docs: List[Document]], delimiter:str]; returns str; concatenates the docs with the
+        specified delimiter and returns a string
+    - convert_to_docs: parameters are texts: List[str]; returns List[Document]; converts the input list of strings
+        to a list of Documents
+    """
+
+    outgoing_edges = 1
+
+    def __init__(self, inputs: Dict[str, Any]):
+        """
+        Initializes a Shaper component.
+
+        :param inputs: A dictionary of input parameters for the Shaper component. These directives
+        are a dictionary version of YAML directives that specify the functions to invoke on the invocation context.
+
+        For example, in the Python snippet below:
+        ```python
+            shaper = Shaper(inputs={"query": {"output": "questions"}})
+            pipeline = Pipeline()
+            pipeline.add_node(component=shaper, name="shaper", inputs=["Query"])
+            ...
+        ```
+        Shaper component is initialized with a directive to rename the invocation context variable query to questions.
+        """
+        super().__init__()
+        self.inputs = inputs
+
+        concat: Callable[[List[str], str], str] = lambda docs, delimiter=" ": delimiter.join(docs)  # type: ignore
```

Comment:
I would implement these callables as free functions or private methods: that way the code would be more readable and we could add tests for them.